frozen funds

ryanberckmans opened the pull request:
https://github.com/AugurProject/augur/issues/988

(frozen funds algorithms in this PR are updated from those mentioned in the ticket)

nuevoalex commented on src/server/getters/get-profit-loss.ts:
```diff
@@ -384,6 +388,9 @@ export async function getProfitLossSummary(db: Knex, augur: Augur, params: GetPr
       outcome: startProfit.outcome,
       marketId: startProfit.marketId,
+      // TODO Alex - are these right?
+      frozenFunds: startProfit.frozenFunds,
```
This should be negated

nuevoalex commented on src/blockchain/log-processors/profit-loss/frozen-funds.ts:
```diff
@@ -0,0 +1,99 @@
+  }
+  // params.event is a trade executed; TODO break each event into a separate function
+  const trade = params.event.tradeData;
+  const frozenFundsAfterEvent = Object.assign({}, params.frozenFundsBeforeEvent);
```
why an object? seems like we only use the `frozenFunds` attribute so why not just have `frozenFunds`?

nuevoalex commented on src/blockchain/log-processors/profit-loss/frozen-funds.ts:
```diff
@@ -0,0 +1,99 @@
+
+  // TODO Alex - can you confirm formula for this:
+  const creatorPortionOfTokensFreedFromEscrow = tokensFreedFromEscrowByThisTrade; // TODO actual value I think is f(minPrice, maxPrice, price)?
+  const fillerPortionOfTokensFreedFromEscrow = tokensFreedFromEscrowByThisTrade; // TODO actual value I think is f(minPrice, maxPrice, price)?
```
Whoever was long paid the exact price of the trade so the way Id structure this is

`
numCompleteSetsSold = BigNumber.min(trade.numCreatorShares, trade.numFillerShares);
longPay = price * numCompleteSetsSold;
shortPay = tokensFreedFromEscrowByThisTrade - longPay;
`

Where filler and creator portions are assigned based on who was long and who was short

nuevoalex commented on src/blockchain/log-processors/profit-loss/frozen-funds.ts:
```diff
@@ -0,0 +1,99 @@
+    // profit and loss for this trade. Ie. frozen profit is profit that I could claim,
+    // but haven't, and we detect this by looking for a non-increase in available funds.
+    // TODO Alex - is this right?
+    frozenFundsAfterEvent.frozenProfit = frozenFundsAfterEvent.frozenProfit.plus(params.event.realizedProfit);
```
frozen profit doesn't need to be stored outside of frozen funds, its just a part of calculating the frozen funds and wont be displayed or used in its own right. This should just be `frozenFunds += params.event.realizedProfit`

frozen profit is profit or loss which is not available in the users funds as it only applied in the context of adjusting the entry into a new position. It's not profit that you could claim but haven't, that's _unrealized_ profit.

nuevoalex commented on src/blockchain/log-processors/profit-loss/update-profit-loss.ts:
```diff
@@ -98,6 +104,15 @@ export async function updateProfitLoss(db: Knex, marketId: Address, positionDelt
+      frozenFunds: oldFrozenFunds,
+      frozenProfit: oldFrozenProfit,
+    },
+    // TODO Alex - is this right profit from "frozenProfit += available_funds decreased ? realized PL : 0"?
```
This isn't right. frozen profit shouldn't be stored or incremented over time. its just part of calculating frozen funds

nuevoalex commented on src/migrations/20190201155650_wcl_profit_loss_timeseries.ts:
```diff
@@ -26,18 +30,36 @@ exports.up = async (knex: Knex): Promise<any> => {
     table.string("position", 42).notNullable();
     table.string("profit", 255).defaultTo("0");
+    table.string("frozenFunds", 255).defaultTo("0");
+    table.string("frozenProfit", 255).defaultTo("0");
```
Frozen profit should not be stored in the DB

nuevoalex commented on src/blockchain/log-processors/profit-loss/frozen-funds.ts:
```diff
@@ -0,0 +1,99 @@
+    // TODO doc formulas
+    frozenFundsAfterEvent.frozenFunds = frozenFundsAfterEvent.frozenFunds.minus(creatorPortionOfTokensFreedFromEscrow);
+    frozenFundsAfterEvent.frozenFunds = frozenFundsAfterEvent.frozenFunds.plus(trade.numCreatorTokens);
+    frozenFundsAfterEvent.frozenFunds = frozenFundsAfterEvent.frozenFunds.minus(trade.numFillerTokens);
```
Realizing now this won't quite work since the goal is to figure out how many tokens the traders actually received and this won't always coincide with that.

I think the best way may be to actually look at shares provided and the price.

So For tokens spent you can still use `frozenFunds = .frozenFunds.plus(trade.numCreatorTokens);`

but then we can avoid the complete set calc entirely and the above line and instead just do something like:

`tokensReceived = trade.numCreatorShares.multipliedBy(price);
frozenFunds = frozenFunds.minus(tokensReceived);`

where price will be `maxPrice - price` for the party shorting (whichever that is)

ryanberckmans commented:
@nuevoalex ready for review. Per our chat I'll next add test cases from chwy's spreadsheet

nuevoalex commented on src/blockchain/log-processors/profit-loss/frozen-funds.ts:
```diff
@@ -0,0 +1,87 @@
+  frozenFundsAfterEvent = frozenFundsAfterEvent.minus(myTokensReceived);
+  frozenFundsAfterEvent = frozenFundsAfterEvent.plus(myTokensSent);
+
+  if (frozenFundsAfterEvent.isGreaterThanOrEqualTo(
```
As I mentioned I suspect we should actually remove this condition and just apply profit any time there if a non-zero PL realized. Tests should reveal this is correct once they've been added

ryanberckmans commented:
Added all test cases from chwy's WCL sheet to integration and unit tests. Fixed two formula bugs. This is nearly ready to merge, just need to ask chwy about Cat3-Tr3 State7.
